_posts/2016-04-30-akka-basics.md

```diff
@@ -0,0 +1,75 @@
+---
+layout: post
+name: Akka-basics
+title: 'Akka basics'
+date: 2016-04-30 00:20:00 +00:00
+author: Felipe Fernández
+image:
+    src: /assets/img/custom/blog/akka.jpg
+    attribution:
+      text: Akka_20120331 by Helena Whitlock, SGU
+      href: https://www.flickr.com/photos/geologicalsurveyofsweden/8703286181
+tags:
+- akka
+- OOP
+- functional programming
+
+---
+
+As [the previous post](http://codurance.com/2016/04/28/async-systems-with-sync-clients) explained our system is based on Akka. Before going into more detail about our solution, I'd like to explain what Akka is about and why it is so awesome.
+
+Akka is a toolkit and not a framework, you can simply use the bits that you need for your service. In this series we'll focus on the core features and we won't go through Akka Stream, Akka-Http or Akka Cluster (since they're out of scope, not because they're not fantastic).
+
+Akka provides a different abstraction to deal with concurrency, paralellism and fault-tolerance. That abstraction is called Actor Model. If you have experience with old Java ecosystem you will know how hard is [writing safe and correct multithreaded code](http://codurance.com/2015/12/13/testing-multithreaded-code-in-java/). Providing human-friendly abstractions like Actors it's a revolution similar to what Java did with memory management. New challenges derived from internet scale and cloud computing requires a reactive approach to programming. Akka aligns with that philosophy and enables you to implement [Event-driven architectures](https://www.wikiwand.com/en/Event-driven_architecture).
+
+## Actors all the way
+
+Threads are an expensive resource, hence we need to use them judiciously. Blocking threads while we're waiting for some I/O operation to respond is really inefficient. Actors use threads in a different way, as a result of which they're pretty lightweight ([several million actors per GB of heap memory](http://doc.akka.io/docs/akka/2.4.4/intro/what-is-akka.html))
+
+Actors mix Object Oriented and Functional Programming principles. OOP is basically about messages as Alan Kay [stated](http://lists.squeakfoundation.org/pipermail/squeak-dev/1998-October/017019.html)
```

Comment:
Actors are a good blend of Object Oriented and Functional Programming principles - if you like to say this otherwise stick to the one you have.